presto-main/src/main/java/com/facebook/presto/type/Re2JRegexp.java

```diff
@@ -0,0 +1,160 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.type;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.google.re2j.Matcher;
+import com.google.re2j.Options;
+import com.google.re2j.Pattern;
+import io.airlift.log.Logger;
+import io.airlift.slice.Slice;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.primitives.Ints.checkedCast;
+import static com.google.re2j.Options.Algorithm.DFA_FALLBACK_TO_NFA;
+import static java.lang.String.format;
+
+public final class Re2JRegexp
+{
+    private static final Logger log = Logger.get(Re2JRegexp.class);
+
+    private static final java.util.regex.Pattern DOT_STAR_PREFIX_SUFFIX_PATTERN = java.util.regex.Pattern.compile("^(\\.\\*\\??)?(.*?)(\\.\\*\\??)?$");
+    private static final int CORE_PATTERN_INDEX = 2;
+
+    public final int dfaStatesLimit;
+    public final int dfaRetries;
+
+    public final Pattern re2jPattern;
+    public final Pattern re2jPatternWithoutDotStartPrefixSuffix;
+
+    public Re2JRegexp(int dfaStatesLimit, int dfaRetries, Slice pattern)
+    {
+        this.dfaStatesLimit = dfaStatesLimit;
+        this.dfaRetries = dfaRetries;
+
+        Options options = Options.builder()
+                .setAlgorithm(DFA_FALLBACK_TO_NFA)
+                .setMaximumNumberOfDFAStates(dfaStatesLimit)
+                .setNumberOfDFARetries(dfaRetries)
+                .setEventsListener(new RE2JEventsListener())
+                .build();
+
+        String patternString = pattern.toStringUtf8();
+        re2jPattern = Pattern.compile(patternString, options);
+
+        // Remove .* prefix and suffix. This will give performance boost when using RE2J Pattern.find() function.
+        java.util.regex.Matcher dotStarPrefixSuffixMatcher = DOT_STAR_PREFIX_SUFFIX_PATTERN.matcher(patternString);
+        checkState(dotStarPrefixSuffixMatcher.matches());
+        String patternStringWithoutDotStartPrefixSuffix = dotStarPrefixSuffixMatcher.group(CORE_PATTERN_INDEX);
+
+        if (!patternStringWithoutDotStartPrefixSuffix.equals(patternString)) {
+            re2jPatternWithoutDotStartPrefixSuffix = Pattern.compile(patternStringWithoutDotStartPrefixSuffix, options);
+        }
+        else {
+            re2jPatternWithoutDotStartPrefixSuffix = re2jPattern;
+        }
+    }
+
+    public boolean matches(Slice source)
+    {
+        return re2jPatternWithoutDotStartPrefixSuffix.find(source);
```

Comment:
I don't agree with you that RE2 is not meant to optimize user suboptimal queries. From what I understand, RE2 was originally designed to take inputs from arbitrary users (not programmers) on the Internet. But anyways, you are the expert here on this subject. And I trust you this is not well optimized in RE2.

I'll merge this within this week. In the long run, we still hope that this can be provided by the engine, and not by Presto. Stripping regular expression components without actually parsing and understanding the pattern string simply feels horrible. Maybe it can work if an option is provided when constructing the pattern. But I don't really know whether this is feasible and/or desirable. I would understand you if you believe adding the option is the worst idea ever. You can probably come up with something better.

Sounds good to you?